ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce/device/all_reduce_op.cpp

```diff
@@ -0,0 +1,138 @@
+    const Tensor& input_tensor,
+    ttnn::operations::reduction::ReduceType math_op,
+    const uint32_t num_links,
+    const MemoryConfig& output_mem_config,
+    ttnn::ccl::Topology topology,
+    const std::optional<size_t> user_defined_num_workers,
+    const std::optional<size_t> user_defined_num_buffers_per_channel) {
+    ttnn::operations::binary::BinaryOpType binary_op_type = convert_reduce_type_to_eltwise_type(math_op);
+    TT_FATAL(std::getenv("TT_METAL_SLOW_DISPATCH_MODE") == nullptr, "This op is only supported for Fast Dispatch");
+    TT_FATAL(topology == ttnn::ccl::Topology::Ring, "All Reduce op is currently supported only on Ring topology");
+
+    auto devices = input_tensor.get_workers();
+    std::vector<Tensor> output_tensors = {Tensor(operation::get_workers_for_op_output({input_tensor}))};
+    operation::launch_op(
+        [binary_op_type, num_links, output_mem_config, topology, devices, user_defined_num_workers, user_defined_num_buffers_per_channel](
+            const std::vector<Tensor>& input_tensors,
+            const std::vector<std::optional<const Tensor>>& optional_input_tensors,
+            const std::vector<std::optional<Tensor>>& optional_output_tensors) mutable -> std::vector<Tensor> {
+            TT_FATAL(input_tensors.size() >= 1, "All reduce op expects an input tensor but it received none");
+            bool is_linear = topology == ttnn::ccl::Topology::Linear;
+
+            const auto& input_tensor = input_tensors.at(0);
+            uint32_t num_devices = devices.size();
+            uint32_t device_index = 0; // Initialize device index
+            std::optional<chip_id_t> receiver_device_id = std::nullopt; // Initialize receiver device ID
+            std::optional<chip_id_t> sender_device_id = std::nullopt; // Initialize sender device ID
+            for (uint32_t i = 0; i < num_devices; ++i) {
+                if (devices.at(i) == input_tensor.device()) {
+
+                    bool is_last_chip_in_clockwise_direction = is_linear && i == (num_devices - 1);
+                    bool is_last_chip_in_counter_clockwise_direction = is_linear && i == 0;
+                    device_index = i;
+                    receiver_device_id = is_last_chip_in_clockwise_direction ?
+                        std::nullopt :
+                        std::optional<chip_id_t>(devices.at((i + 1) % num_devices)->id());
+                    sender_device_id = is_last_chip_in_counter_clockwise_direction ?
+                        std::nullopt :
+                        std::optional<chip_id_t>(devices.at((i + num_devices - 1) % num_devices)->id());
+                    break;
+                }
+            }
+            TT_FATAL(receiver_device_id != std::nullopt || sender_device_id != std::nullopt, "Error in all reduce op setup");
+
+            auto shape = input_tensor.shape();
+            auto rank = shape.rank();
+
+            uint32_t merged_dim_size = 1;
+            for (uint32_t i = 0; i <= rank - 3; ++i) {
```

Comment:
should we check whether rank >= 3? 
Another option might be `for (uint32_t i = 0; i + 3 <= rank; ++i)`

Reply:
Yes, I will add a check before this

Reply:
We want to merge all of the outer or batch dims (0 to n-3 inclusive) so we can insert a new dimension and sum along that new other dimension. Though that being said there is a bug here for rank < 2 and for that we should fix how `new_shape` is calculated.

Reply:
Oh I see what you were referring to @rfurko-tt - `rank` is unsigned. Then disregard the first part of my comment as I thought you were alluding to something else. The latter part still applies here.